doc/proxy-settings.md

# ProxySettings

Represents the proxy server configurations for API calls.

## Properties

| Name | Type | Tag | Description |
|  --- | --- | --- | --- |
| address | `str` | required | The proxy server URL. |
| port | `int` | optional | The port to connect to the proxy server. |
| username | `str` | optional | Username for proxy authentication. |
| password | `str` | optional | Password for proxy authentication. |

## Usage Example

```python
from multiauthsample.multiauthsample_client import MultiauthsampleClient
from multiauthsample.http.proxy_settings import ProxySettings

client = MultiauthsampleClient(
    proxy_settings=ProxySettings(
        address='http://localhost',
        port=8888,
        username='user',
        password='pass'
     )
)
```



doc/environment-based-client-initialization.md

# Environment-Based Client Initialization

The SDK client can also be initialized directly from environment variables using the `from_environment()` class method. This allows the SDK to automatically read configuration values from the runtime environment or a .env file.

## Example

```python
from multiauthsample.multiauthsample_client import MultiauthsampleClient

# Specify the path to your .env file if it’s located outside the project’s root directory.
client = MultiauthsampleClient.from_environment(dotenv_path='/path/to/.env')
```

You can also specify a path to a `.env` file by passing it to the `from_environment()` method:

The same method can accept keyword arguments to override any values read from the environment, and the arguments to override values should follow the same approach as code-based client initialization.

```python
from multiauthsample.multiauthsample_client import MultiauthsampleClient

client = MultiauthsampleClient.from_environment(
    dotenv_path='/path/to/.env',
    timeout=0,  # overrides timeout from environment variable
)
```

Values provided through arguments take precedence over those defined in environment variables.

## Example `.env` File

```python
PORT=80
SUITES=1
ACCESS_TOKEN=
ENVIRONMENT=testing

BASIC_AUTH_USERNAME=ExampleUsername
BASIC_AUTH_PASSWORD=ExamplePassword
API_KEY_TOKEN=
API_KEY_API_KEY=
API_HEADER_TOKEN=
API_HEADER_API_KEY=
O_AUTH_CCG_O_AUTH_CLIENT_ID=ExampleOAuthClientId
O_AUTH_CCG_O_AUTH_CLIENT_SECRET=ExampleOAuthClientSecret
O_AUTH_CCG_O_AUTH_CLOCK_SKEW=5
O_AUTH_ACG_O_AUTH_CLIENT_ID=ExampleOAuthClientId
O_AUTH_ACG_O_AUTH_CLIENT_SECRET=ExampleOAuthClientSecret
O_AUTH_ACG_O_AUTH_REDIRECT_URI=ExampleOAuthRedirectUri
O_AUTH_ACG_O_AUTH_SCOPES=ExampleOAuthScopes
O_AUTH_ROPCG_O_AUTH_CLIENT_ID=ExampleOAuthClientId
O_AUTH_ROPCG_O_AUTH_CLIENT_SECRET=ExampleOAuthClientSecret
O_AUTH_ROPCG_O_AUTH_USERNAME=ExampleOAuthUsername
O_AUTH_ROPCG_O_AUTH_PASSWORD=ExampleOAuthPassword
O_AUTH_BEARER_TOKEN_ACCESS_TOKEN=ExampleAccessToken

TIMEOUT=60
MAX_RETRIES=3
BACKOFF_FACTOR=2
RETRY_STATUSES=408,413
RETRY_METHODS=GET,PUT,DELETE

# Proxy Configuration
PROXY_ADDRESS=http://localhost:3000
PROXY_PORT=8080
PROXY_USERNAME=username
PROXY_PASSWORD=password
```

## Note

- If an environment variable is not defined, the default SDK configuration value will be used.



multiauthsample/configuration.py
"""multiauthsample.

This file was automatically generated by APIMATIC v3.0 ( https://www.apimatic.io ).
"""

# ruff: noqa: E501
import os
from enum import Enum

from apimatic_core.http.configurations.http_client_configuration import (
    HttpClientConfiguration,
)
from apimatic_requests_client_adapter.requests_client import (
    RequestsClient,
)
from dotenv import load_dotenv

from multiauthsample.api_helper import APIHelper
from multiauthsample.http.proxy_settings import (
    ProxySettings,
)
from multiauthsample.models.suite_code_enum import (
    SuiteCodeEnum,
)


class Environment(Enum):
    """An enum for SDK environments."""

    PRODUCTION = 0
    TESTING = 1

    @classmethod
    def from_value(cls, value, default=None):
        """Convert a value (string or int) to an Environment enum member.

        Args:
            value (Union[str, int]): The value to convert.
            default (Environment): The fallback enum member if conversion fails.

        Returns:
            Environment: Matching enum member or fallback if invalid.

        """
        if value is None:
            return default

        # Try to match directly by enum member
        if isinstance(value, cls):
            return value

        # Handle integer or string conversion
        for member in cls:
            if (str(member.value).lower() == str(value).lower()
                or member.name.lower() == str(value).lower()):
                return member

        # Fallback to provided default
        return default


class Server(Enum):
    """An enum for API servers."""

    DEFAULT = 0
    AUTH = 1

    @classmethod
    def from_value(cls, value, default=None):
        """Convert a value (string or int) to a Server enum member.

        Args:
            value (Union[str, int]): The value to convert.
            default (Server): The fallback enum member if conversion fails.

        Returns:
            Server: Matching enum member or fallback if invalid.

        """
        if value is None:
            return default

        # Try to match directly by enum member
        if isinstance(value, cls):
            return value

        # Handle integer or string conversion
        for member in cls:
            if (str(member.value).lower() == str(value).lower()
                or member.name.lower() == str(value).lower()):
                return member

        # Fallback to provided default
        return default


class Configuration(HttpClientConfiguration):
    """A class used for configuring the SDK by a user."""

    @property
    def environment(self):
        """Return current environment."""
        return self._environment

    @property
    def port(self):
        """Return current port."""
        return self._port

    @property
    def suites(self):
        """Return current suites."""
        return self._suites

    @property
    def basic_auth_credentials(self):
        """Return current basic_auth_credentials."""
        return self._basic_auth_credentials

    @property
    def api_key_credentials(self):
        """Return current api_key_credentials."""
        return self._api_key_credentials

    @property
    def api_header_credentials(self):
        """Return current api_header_credentials."""
        return self._api_header_credentials

    @property
    def o_auth_ccg_credentials(self):
        """Return current o_auth_ccg_credentials."""
        return self._o_auth_ccg_credentials

    @property
    def o_auth_acg_credentials(self):
        """Return current o_auth_acg_credentials."""
        return self._o_auth_acg_credentials

    @property
    def o_auth_ropcg_credentials(self):
        """Return current o_auth_ropcg_credentials."""
        return self._o_auth_ropcg_credentials

    @property
    def o_auth_bearer_token_credentials(self):
        """Return current o_auth_bearer_token_credentials."""
        return self._o_auth_bearer_token_credentials

    @property
    def custom_auth_credentials(self):
        """Return current custom_auth_credentials."""
        return self._custom_auth_credentials

    @property
    def access_token(self):
        """Return current access_token."""
        return self._access_token

    def __init__(self, http_client_instance=None,
                 override_http_client_configuration=False, http_call_back=None,
                 timeout=60, max_retries=0, backoff_factor=2, retry_statuses=None,
                 retry_methods=None, proxy_settings=None,
                 environment=Environment.TESTING, port="80", suites=1,
                 basic_auth_credentials=None, api_key_credentials=None,
                 api_header_credentials=None, o_auth_ccg_credentials=None,
                 o_auth_acg_credentials=None, o_auth_ropcg_credentials=None,
                 o_auth_bearer_token_credentials=None, custom_auth_credentials=None,
                 access_token=""):
        """Initialize Configuration object."""
        if retry_methods is None:
            retry_methods = ["GET", "PUT"]

        if retry_statuses is None:
            retry_statuses = [408, 413, 429, 500, 502, 503, 504, 521,
                522, 524]

        super().__init__(
            http_client_instance=http_client_instance,
            override_http_client_configuration=override_http_client_configuration,
            http_call_back=http_call_back, timeout=timeout,
            max_retries=max_retries, backoff_factor=backoff_factor,
            retry_statuses=retry_statuses, retry_methods=retry_methods,
            proxy_settings=proxy_settings,
        )

        # Current API environment
        self._environment = environment

        # port value
        self._port = port

        # suites value
        self._suites = suites

        # The object holding Basic Authentication credentials
        self._basic_auth_credentials = basic_auth_credentials

        # The object holding Custom Query Parameter credentials
        self._api_key_credentials = api_key_credentials

        # The object holding Custom Header Signature credentials
        self._api_header_credentials = api_header_credentials

        # The object holding OAuth 2 Client Credentials Grant credentials
        self._o_auth_ccg_credentials = o_auth_ccg_credentials

        # The object holding OAuth 2 Authorization Code Grant credentials
        self._o_auth_acg_credentials = o_auth_acg_credentials

        # The object holding OAuth 2 Resource Owner Credentials Grant credentials
        self._o_auth_ropcg_credentials = o_auth_ropcg_credentials

        # The object holding OAuth 2 Bearer token credentials
        self._o_auth_bearer_token_credentials = o_auth_bearer_token_credentials

        # The object holding Custom Authentication credentials
        self._custom_auth_credentials = custom_auth_credentials

        # TODO: Replace
        self._access_token = access_token

        # The Http Client to use for making requests.
        self.set_http_client(self.create_http_client())

    def clone_with(self, http_client_instance=None,
                   override_http_client_configuration=None, http_call_back=None,
                   timeout=None, max_retries=None, backoff_factor=None,
                   retry_statuses=None, retry_methods=None, proxy_settings=None,
                   environment=None, port=None, suites=None,
                   basic_auth_credentials=None, api_key_credentials=None,
                   api_header_credentials=None, o_auth_ccg_credentials=None,
                   o_auth_acg_credentials=None, o_auth_ropcg_credentials=None,
                   o_auth_bearer_token_credentials=None, custom_auth_credentials=None,
                   access_token=None):
        """Clone configuration with overrides."""
        http_client_instance = http_client_instance or self.http_client_instance
        override_http_client_configuration =\
            (override_http_client_configuration
            or self.override_http_client_configuration)
        http_call_back = http_call_back or self.http_callback
        timeout = timeout or self.timeout
        max_retries = max_retries or self.max_retries
        backoff_factor = backoff_factor or self.backoff_factor
        retry_statuses = retry_statuses or self.retry_statuses
        retry_methods = retry_methods or self.retry_methods
        proxy_settings = proxy_settings or self.proxy_settings
        environment = environment or self.environment
        port = port or self.port
        suites = suites or self.suites
        basic_auth_credentials = basic_auth_credentials or self.basic_auth_credentials
        api_key_credentials = api_key_credentials or self.api_key_credentials
        api_header_credentials = api_header_credentials or self.api_header_credentials
        o_auth_ccg_credentials = o_auth_ccg_credentials or self.o_auth_ccg_credentials
        o_auth_acg_credentials = o_auth_acg_credentials or self.o_auth_acg_credentials
        o_auth_ropcg_credentials =\
            (o_auth_ropcg_credentials
            or self.o_auth_ropcg_credentials)
        o_auth_bearer_token_credentials =\
            (o_auth_bearer_token_credentials
            or self.o_auth_bearer_token_credentials)
        custom_auth_credentials =\
            (custom_auth_credentials
            or self.custom_auth_credentials)
        access_token = access_token or self.access_token
        return Configuration(
            http_client_instance=http_client_instance,
            override_http_client_configuration=override_http_client_configuration,
            http_call_back=http_call_back, timeout=timeout, max_retries=max_retries,
            backoff_factor=backoff_factor, retry_statuses=retry_statuses,
            retry_methods=retry_methods, proxy_settings=proxy_settings,
            environment=environment, port=port, suites=suites,
            basic_auth_credentials=basic_auth_credentials,
            api_key_credentials=api_key_credentials,
            api_header_credentials=api_header_credentials,
            o_auth_ccg_credentials=o_auth_ccg_credentials,
            o_auth_acg_credentials=o_auth_acg_credentials,
            o_auth_ropcg_credentials=o_auth_ropcg_credentials,
            o_auth_bearer_token_credentials=o_auth_bearer_token_credentials,
            custom_auth_credentials=custom_auth_credentials, access_token=access_token,
        )

    def create_http_client(self):
        """Create the HTTP client instance."""
        return RequestsClient(
            timeout=self.timeout, max_retries=self.max_retries,
            backoff_factor=self.backoff_factor, retry_statuses=self.retry_statuses,
            retry_methods=self.retry_methods,
            http_client_instance=self.http_client_instance,
            override_http_client_configuration=self.override_http_client_configuration,
            response_factory=self.http_response_factory,
            proxies=self.proxy_settings.to_proxies() if self.proxy_settings else None,
        )

    # All the environments the SDK can run in
    environments = {
        Environment.PRODUCTION: {
            Server.DEFAULT: "http://apimatic.hopto.org:{suites}",
            Server.AUTH: "http://apimaticauth.hopto.org:3000",
        },
        Environment.TESTING: {
            Server.DEFAULT: "http://localhost:3000",
            Server.AUTH: "http://localhost:3000/oauth2/auth-server",
        },
    }

    def get_base_uri(self, server=Server.DEFAULT):
        """Generate the appropriate base URI for the environment and the server.

        Args:
            server (Configuration.Server): The server enum for which the base
            URI is required.

        Returns:
            String: The base URI.

        """
        parameters = {
            "port": {"value": self.port, "encode": False},
            "suites": {"value": self.suites, "encode": False},
        }

        return APIHelper.append_url_with_template_parameters(
            self.environments[self.environment][server], parameters,
        )

    @classmethod
    def from_environment(cls, dotenv_path=None, **overrides):
        """Create Configuration object from .env and environment variables.

        Args:
            dotenv_path (str, optional): Path to the .env file to load.
             If None, the default .env file is used.
            **overrides: Optional values overriding setting from environment variables.

        Returns:
            Configuration: A configuration object populated with the resolved values.

        """
        # load .env automatically
        load_dotenv(dotenv_path or None, override=True)

        override_http_client_configuration = os.getenv(
            "OVERRIDE_HTTP_CLIENT_CONFIGURATION", "false").lower() == "true"
        timeout = int(os.getenv("TIMEOUT", "60"))
        max_retries = int(os.getenv("MAX_RETRIES", "0"))
        backoff_factor = float(os.getenv("BACKOFF_FACTOR", "2"))
        statuses = os.getenv("RETRY_STATUSES", None)
        retry_statuses = [int(v.strip()) for v in statuses.split(",")
                          if v.strip().isdigit()] if statuses else None
        methods = os.getenv("RETRY_METHODS", None)
        retry_methods = [v.strip() for v in methods.split(",") if v.strip()]\
            if methods else None
        environment = Environment.from_value(
            os.getenv("ENVIRONMENT"), Environment.TESTING)
        port = os.getenv("PORT", "80")
        suites = SuiteCodeEnum.from_value(os.getenv("SUITES"), 1)
        access_token = os.getenv("ACCESS_TOKEN", "")

        from multiauthsample.http.auth.api_header import (
            ApiHeaderCredentials,
        )
        from multiauthsample.http.auth.api_key import (
            ApiKeyCredentials,
        )
        from multiauthsample.http.auth.basic_auth import (
            BasicAuthCredentials,
        )
        from multiauthsample.http.auth.o_auth_acg import (
            OAuthACGCredentials,
        )
        from multiauthsample.http.auth.o_auth_bearer_token import (
            OAuthBearerTokenCredentials,
        )
        from multiauthsample.http.auth.o_auth_ccg import (
            OAuthCCGCredentials,
        )
        from multiauthsample.http.auth.o_auth_ropcg import (
            OAuthROPCGCredentials,
        )

        # Preparing default configuration
        default_config = cls(
            override_http_client_configuration=override_http_client_configuration,
            timeout=timeout,
            max_retries=max_retries,
            backoff_factor=backoff_factor,
            retry_statuses=retry_statuses,
            retry_methods=retry_methods,
            environment=environment,
            port=port,
            suites=suites,
            access_token=access_token,
            proxy_settings=ProxySettings.from_environment(),
            basic_auth_credentials=BasicAuthCredentials.from_environment(),
            api_key_credentials=ApiKeyCredentials.from_environment(),
            api_header_credentials=ApiHeaderCredentials.from_environment(),
            o_auth_ccg_credentials=OAuthCCGCredentials.from_environment(),
            o_auth_acg_credentials=OAuthACGCredentials.from_environment(),
            o_auth_ropcg_credentials=OAuthROPCGCredentials.from_environment(),
            o_auth_bearer_token_credentials=OAuthBearerTokenCredentials.from_environment(),
        )

        return default_config.clone_with(**overrides)
